### PR TITLE
[oneDNN] fuse gelu subgraph with MatMul and Biasadd

### PR DESCRIPTION
**Pull request by gaurides**
This PR fuses gelu subgraph with MatMul and BiasAdd. This pattern is found in official bert base, bert large etc models. And gives significant performance improvement. This fusion is also enabled for bfloat16.

### CODE REVIEW

**Comment by gaurides** [comment]
Original sub-graph
gelu_subgraph
After fusion
gelu_fused

**Comment by gbaned** [comment]
Hi @ezhulenev Can you please review this PR ? Thank you!

**Comment by gaurides** [comment]
[@elfringham](https://github.com/elfringham) : do you know why ARM CI failed for this PR also?

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 1]
```diff
@@ -1163,6 +1179,38 @@ bool IsMatchedMatMulBiasAddAndGeluExact(
...
+              {
+                {"MatMul", "matmul", NodeStatus::kRemove},
+                {"*", "bias", NodeStatus::kRemain}
+              }
+            },
+            {"Cast|Const", "one_half", NodeStatus::kRemain}
+          }
+        }
+      }
+    };
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1212]: Nit: The brackets are so far apart it's hard to keep track. Let's add comments at the end of each level. Let's also shorten "bias_add_times_square_root_one_half" to "bias_add_x_sqrt_one_half".
```suggestion
  static utils::OpTypePattern* gelu_exact_pattern2 = new utils::OpTypePattern
    {"Mul", "output", NodeStatus::kReplace,
      {
        {"Add|AddV2", "erf_plus_one", NodeStatus::kRemove,
          {
            {"Erf", "erf", NodeStatus::kRemove,
              {
                {"Mul", "bias_add_x_sqrt_one_half", NodeStatus::kRemove,
                  {
                    {"BiasAdd", "bias_add", NodeStatus::kRemove},
                    {"Cast|Const", "sqrt_one_half", NodeStatus::kRemain}
                  }
                }  // Mul: "bias_add_x_sqrt_one_half"
              }
            },  // Erf: "erf"
            {"Cast|Const", "one", NodeStatus::kRemain}
          }
        },  // Add|AddV2: "erf_plus_one"
        {"Mul", "erf_plus_one_times_one_half", NodeStatus::kRemove,
          {
            {"BiasAdd", "bias_add", NodeStatus::kRemove,
              {
                {"MatMul", "matmul", NodeStatus::kRemove},
                {"*", "bias", NodeStatus::kRemain}
              }
            },  // BiasAdd: "bias_add"
            {"Cast|Const", "one_half", NodeStatus::kRemain}
          }
        }  // Mul: "erf_plus_one_times_one_half"
      }
    };  // Mul: "output"
```

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 2]
```diff
@@ -1163,6 +1179,38 @@ bool IsMatchedMatMulBiasAddAndGeluExact(
   using utils::MatchingDirection;
   using utils::NodeStatus;
   // clang-format off
+  static utils::OpTypePattern* gelu_exact_pattern2 = new utils::OpTypePattern
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1182]: Nit: Pattern 2 should be after pattern 1. Please move this to below the definition of `gelu_exact_pattern`.

Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1182]: Please also add a comment drawing the subgraph, i.e.,
```c++
  // Pattern 2:
  //  Cast|Const: 1/sqrt(2)    Cast|Const: 1
  //                  \               \
  //  * --> BiasAdd --> Mul --> Erf --> Add|AddV2 --> Mul
  //      /         \                               /
  // MatMul           -------------------------> Mul
  //                                           / 
  //                                  Cast|Const: 1/2  
```

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 3]
```diff
@@ -1163,6 +1179,38 @@ bool IsMatchedMatMulBiasAddAndGeluExact(
...
+                {"*", "bias", NodeStatus::kRemain}
+              }
+            },
+            {"Cast|Const", "one_half", NodeStatus::kRemain}
+          }
+        }
+      }
+    };
+
   static utils::OpTypePattern* gelu_exact_pattern = new utils::OpTypePattern
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1214]: Please add a comment drawing the subgraph, i.e.,
```c++
  // Pattern 1:
  //    Const: 1/sqrt(2)        Const: 1    Const: 1/2
  //                  \               \         \
  //  * --> BiasAdd --> Mul --> Erf --> AddV2 --> Mul --> Mul
  //      /         \___________________________________/
  // MatMul
```

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 4]
```diff
@@ -1202,10 +1250,20 @@ bool IsMatchedMatMulBiasAddAndGeluExact(
...
+  if (found_gelu_exact) {
+    return true;
+  } else {
+    matched_nodes_map->clear();
+    remove_node_indices->clear();
+  return graph_matcher.GetMatchedNodes(
+        *gelu_exact_pattern2, ctx.nodes_to_preserve, node_view,
+      matched_nodes_map ? matched_nodes_map : &dummy_matched_nodes_map,
+      remove_node_indices ? remove_node_indices : &dummy_remove_node_indices);
+  }
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1266]: Nit: Let's make this more concise:
```suggestion
  if (!matched_nodes_map) matched_nodes_map = &dummy_matched_nodes_map;
  if (!remove_node_indices) remove_node_indices = &dummy_remove_node_indices;
  if (graph_matcher.GetMatchedNodes(*gelu_exact_pattern, ctx.nodes_to_preserve,
                                    node_view, matched_nodes_map,
                                    remove_node_indices)) {
    return true;
  }
  matched_nodes_map->clear();
  remove_node_indices->clear();
  return graph_matcher.GetMatchedNodes(*gelu_exact_pattern2,
                                       ctx.nodes_to_preserve, node_view,
                                       matched_nodes_map, remove_node_indices);
```

File: tensorflow/python/grappler/remapper_test.py [hunk 5]
```diff
@@ -158,6 +158,10 @@ def _VerifyValues(self, model_fn, use_low_precision, fused_op, epilog_ops):
   @test_util.disable_xla('This test does not pass with XLA')
   def test_matmul_biasadd_activation_fusion(self, mode):
     """Test MatMul+BiasAdd+Gelu fusion."""
+
+
+
+
```
Comment (penpornk) [inline tensorflow/python/grappler/remapper_test.py:164]: Please remove extra blank lines.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the long delay!

File: tensorflow/core/grappler/optimizers/remapper.cc (hunk 3 above)
Comment (gaurides) [reply tensorflow/core/grappler/optimizers/remapper.cc:1178]: Thanks! this is very useful.  Added in the file

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 6]
```diff
@@ -1125,34 +1125,50 @@ inline bool VerifyConstants(RemapperContext* ctx,
...
+      const auto* const_node = regular_node_view->node();
+
+      ret = verify_constant(const_node, it->second);
+
+    } else {
+      ret = verify_constant(node_def, it->second);
     }
   }
-  return true;
+  return ret;
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Thanks, yes, this simplifies and we don't need the lambda function

File: tensorflow/core/grappler/optimizers/mkl_remapper_test.cc [hunk 7]
```diff
@@ -668,20 +668,37 @@ class FusedMatMulBiasAddAndGeluTest : public GrapplerTest {
...
+                                          square_root_one_half_const, DTYPE);
     auto bias_add_times_square_root_one_half =
         ops::Mul(s.WithOpName("bias_add_times_square_root_one_half"), bias_add,
                  square_root_one_half);
     auto erf =
         ops::Erf(s.WithOpName("erf"), bias_add_times_square_root_one_half);
-    auto one = ops::Const(s.WithOpName("one"), {1.0f}, {});
+
+    auto one_const = ops::Const(s.WithOpName("one_const"), {1.0f}, {});
+    auto one = ops::Cast(s.WithOpName("one"), one_const, DTYPE);
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/core/grappler/optimizers/mkl_remapper_test.cc:682]: I added the Cast as this how the pattern shows up in the models graph for BF16. Also, I keep running into errors with Const<T> way.

**Comment by gaurides** [comment]
@penpornk : thanks for the review and the recommendations. I did the review rework for all except 1 change for Cast in the test. Also resolved the merge conflict.

**Comment by gaurides** [comment]
Adding @ezhulenev back to the review, not sure how he was removed

File: tensorflow/core/grappler/optimizers/mkl_remapper_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/grappler/optimizers/mkl_remapper_test.cc:682]: Thank you for the explanation! Would you mind adding a comment in the code saying that sometimes there is a Cast?

**Review by penpornk — APPROVED** [review]
Thank you very much!

**Pull request merged**